File: generated_py/tests/test_parsegen_code.py
"""Unit tests for parser codegen: ensure hint 'type' is applied for with_adopted_grandchildren."""
from __future__ import annotations

import json
import sys
import tempfile
import unittest
from pathlib import Path

# Add generated_py so we can import codegen and runtime
sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from codegen.parsegen_code import main


class TestParsegenCode(unittest.TestCase):
    """Parser codegen tests."""

    def test_with_adopted_grandchildren_respects_type_in_generated_code(self) -> None:
        """Generated parser must apply hint 'type' in with_adopted_grandchildren branch."""
        tables = {
            "start_symbol": "Root",
            "actions": {"0": {"EOF": {"type": "accept"}}},
            "gotos": {},
            "productions": [
                {
                    "lhs": "Root",
                    "rhs": [
                        {"name": "lbracket", "terminal": True},
                        {"name": "Elements", "terminal": False},
                        {"name": "rbracket", "terminal": True},
                    ],
                    "hint": {
                        "parent_literal": "[]",
                        "with_adopted_grandchildren": [1],
                        "type": "array",
                    },
                },
            ],
            "hint_mode": "hints",
        }
        with tempfile.NamedTemporaryFile(
            mode="w", suffix=".json", delete=False
        ) as f:
            json.dump(tables, f)
            json_path = Path(f.name)
        try:
            out_path = Path(tempfile.gettempdir()) / "test_array_parser_py_pgpg.py"
            argv = [
                "parsegen_code.py",
                "-o", str(out_path),
                "-c", "ArrayParser",
                "--prefix", "",
                str(json_path),
            ]
            old_argv = sys.argv
            sys.argv = argv
            try:
                self.assertEqual(main(), 0)
            finally:
                sys.argv = old_argv
            code = out_path.read_text()
            self.assertIn(
                "node_type = prod.node_type or parent_type",
                code,
                "generated code should apply hint type in with_adopted_grandchildren branch",
            )
            self.assertTrue(
                'node_type="array"' in code or "node_type='array'" in code,
                "generated production table should include node_type for the hint",
            )
            self.assertIn(
                "new_ast_node(parent_token, node_type, new_children)",
                code,
                "generated code should build node with node_type in with_adopted_grandchildren branch",
            )
        finally:
            json_path.unlink(missing_ok=True)
            out_path.unlink(missing_ok=True)


if __name__ == "__main__":
    unittest.main()

File: generator_js/tests/test_parsegen_code.js
/**
 * Unit tests for parser codegen: ensure hint 'type' is applied for with_adopted_grandchildren.
 */
import fs from "fs";
import os from "os";
import path from "path";
import { describe, it } from "node:test";
import assert from "node:assert";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const codegenDir = path.join(__dirname, "..", "codegen");

async function runParsegenCode(outputPath, jsonPath, className, prefix = "") {
  const scriptPath = path.join(codegenDir, "parsegen_code.js");
  const args = [
    "-o", outputPath,
    "-c", className,
    "--prefix", prefix,
    jsonPath,
  ];
  const { spawn } = await import("node:child_process");
  const proc = spawn("node", [scriptPath, ...args], {
    stdio: "pipe",
    cwd: path.join(__dirname, ".."),
  });
  return new Promise((resolve) => {
    let err = "";
    proc.stderr?.on("data", (c) => (err += c));
    proc.on("close", (code) => resolve({ code, err }));
  });
}

describe("parsegen_code", () => {
  it("generated parser applies hint type in with_adopted_grandchildren branch", async () => {
    const tmp = os.tmpdir();
    const jsonPath = path.join(tmp, "pgpg_js_test_parsegen_" + Date.now() + ".json");
    const outPath = path.join(tmp, "pgpg_js_test_array_parser_" + Date.now() + ".js");
    const tables = {
      start_symbol: "Root",
      actions: { "0": { EOF: { type: "accept" } } },
      gotos: {},
      productions: [
        {
          lhs: "Root",
          rhs: [
            { name: "lbracket", terminal: true },
            { name: "Elements", terminal: false },
            { name: "rbracket", terminal: true },
          ],
          hint: {
            parent_literal: "[]",
            with_adopted_grandchildren: [1],
            type: "array",
          },
        },
      ],
      hint_mode: "hints",
    };
    fs.writeFileSync(jsonPath, JSON.stringify(tables), "utf-8");
    try {
      const { code } = await runParsegenCode(outPath, jsonPath, "ArrayParser", "");
      assert.strictEqual(code, 0, "parsegen_code should exit 0");
      const codeStr = fs.readFileSync(outPath, "utf-8");
      assert.ok(
        codeStr.includes("nodeType = prod.nodeType || parentType"),
        "generated code should apply hint type in with_adopted_grandchildren branch",
      );
      assert.ok(
        codeStr.includes('nodeType: "array"'),
        "generated production table should include node_type for the hint",
      );
      assert.ok(
        codeStr.includes("newASTNode(parentToken, nodeType, newChildren)"),
        "generated code should build node with nodeType in with_adopted_grandchildren branch",
      );
    } finally {
      try {
        fs.unlinkSync(jsonPath);
        fs.unlinkSync(outPath);
      } catch (_) {}
    }
  });
});
